src/Dependencies/PooledObjects/PooledDelegates.cs

```diff
@@ -0,0 +1,404 @@
+        /// argument:</para>
+        ///
+        /// <code>
+        /// int x = 3;
+        /// RunWithPredicate((a, b, c) => this.IsSomething(a, b, c, x));
+        /// </code>
+        ///
+        /// <para>The following example shows the use of a pooled delegate to avoid capturing allocations for the same
+        /// predicate:</para>
+        ///
+        /// <code>
+        /// int x = 3;
+        /// using var _ = GetPooledFunction((a, b, c, arg) => arg.self.IsSomething(a, b, c, arg.x), (self: this, x), out Func&lt;int, int, int, bool&gt; predicate);
+        /// RunWithPredicate(predicate);
+        /// </code>
+        /// </example>
+        /// <typeparam name="T1">The type of the first parameter of the bound function.</typeparam>
+        /// <typeparam name="T2">The type of the second parameter of the bound function.</typeparam>
+        /// <typeparam name="T3">The type of the third parameter of the bound function.</typeparam>
+        /// <typeparam name="TArg">The type of argument to pass to <paramref name="unboundFunction"/>.</typeparam>
+        /// <typeparam name="TResult">The type of the return value of the function.</typeparam>
+        /// <param name="unboundFunction">The unbound function delegate.</param>
+        /// <param name="argument">The argument to pass to the unbound function delegate.</param>
+        /// <param name="boundFunction">A delegate which calls <paramref name="unboundFunction"/> with the specified
+        /// <paramref name="argument"/>.</param>
+        /// <returns>A disposable <see cref="Releaser"/> which returns the object to the delegate pool.</returns>
+        public static Releaser GetPooledFunction<T1, T2, T3, TArg, TResult>(Func<T1, T2, T3, TArg, TResult> unboundFunction, TArg argument, out Func<T1, T2, T3, TResult> boundFunction)
+            => GetPooledDelegate<FuncWithBoundArgument<T1, T2, T3, TArg, TResult>, TArg, Func<T1, T2, T3, TArg, TResult>, Func<T1, T2, T3, TResult>>(unboundFunction, argument, out boundFunction);
+
+        public struct Releaser : IDisposable
+        {
+            private readonly object _pooledObject;
+            private readonly Action<object> _releaser;
+
+            internal Releaser(object pooledObject, Action<object> releaser)
+            {
+                _pooledObject = pooledObject;
+                _releaser = releaser;
+            }
+
+            public void Dispose() => _releaser(_pooledObject);
+        }
+
+        private abstract class AbstractDelegateWithBoundArgument<TArg, TUnboundDelegate, TBoundDelegate>
+            where TUnboundDelegate : Delegate
+            where TBoundDelegate : Delegate
+        {
+            protected AbstractDelegateWithBoundArgument()
+            {
+                BoundDelegate = Bind();
+            }
+
+            public TArg Argument { get; set; }
+            public TUnboundDelegate UnboundDelegate { get; set; }
```

Comment:
>public TUnboundDelegate UnboundDelegate { get; set; } [](start = 12, length = 53)

Do we change these two separately? If not, perhaps we should have the settable only through a special method that takes both together.

Reply:
Refactored to use an `Initialize` method.